get the actual user instead of a federated cloud id

schiessle opened the pull request:
`$view->getUidAndFilename($filename);` returns the federated cloud id in case of a federated share. But in the case of versioning and trashbin we need the local user who "owns" the file which is the current logged in user in case of a federated share.

This reverts changes introduced here by @icewind1991: https://github.com/owncloud/core/commit/300eb54c871cfe48165ee32ecdc5067226aa0b7b 

While in most cases in makes sense to return the federated cloud id of the owner of the file. In this cases it breaks the trashbin and versions.

Steps to test:
- set up two ownClouds
- ownCloud1 shares a folder which contains a file to ownCloud2
- try to delete the file on ownCloud2

Expected: 

File gets deleted and added to the trashbin of owner@ownCloud1 and to the user at ownCloud2

Actual Result: 

delete fails with this exception:

```
{"reqId":"h3FJVmACIo8+mISwiP+g","remoteAddr":"::1","app":"files","message":" Backends provided no user object for schiesbn@localhost\/federation\/","level":3,"time":"2016-02-16T16:08:39+00:00"}
{"reqId":"h3FJVmACIo8+mISwiP+g","remoteAddr":"::1","app":"no app in context","message":"Exception: {\"Exception\":\"OC\\\\User\\\\NoUserException\",\"Message\":\"Backends provided no user object for schiesbn@localhost\\\/federation\\\/\",\"Code\":0,\"Trace\":\"#0 \\\/home\\\/src\\\/owncloud\\\/server\\\/master\\\/apps\\\/files_trashbin\\\/lib\\\/trashbin.php(80): OC\\\\Files\\\\Filesystem::initMountPoints('schiesbn@localh...')\\n#1 \\\/home\\\/src\\\/owncloud\\\/server\\\/master\\\/apps\\\/files_trashbin\\\/lib\\\/trashbin.php(67): OCA\\\\Files_Trashbin\\\\Trashbin::getUidAndFilename('\\\/New folder\\\/wel...')\\n#2 [internal function]: OCA\\\\Files_Trashbin\\\\Trashbin::ensureFileScannedHook(Array)\\n#3 \\\/home\\\/src\\\/owncloud\\\/server\\\/master\\\/lib\\\/private\\\/hook.php(104): call_user_func(Array, Array)\\n#4 \\\/home\\\/src\\\/owncloud\\\/server\\\/master\\\/lib\\\/private\\\/files\\\/view.php(1172): OC_Hook::emit('OC_Filesystem', 'delete', Array)\\n#5 \\\/home\\\/src\\\/owncloud\\\/server\\\/master\\\/lib\\\/private\\\/files\\\/view.php(1048): OC\\\\Files\\\\View->runHooks(Array, '\\\/New folder\\\/wel...')\\n#6 \\\/home\\\/src\\\/owncloud\\\/server\\\/master\\\/lib\\\/private\\\/files\\\/view.php(641): OC\\\\Files\\\\View->basicOperation('unlink', '\\\/New folder\\\/wel...', Array)\\n#7 \\\/home\\\/src\\\/owncloud\\\/server\\\/master\\\/apps\\\/dav\\\/lib\\\/connector\\\/sabre\\\/file.php(320): OC\\\\Files\\\\View->unlink('\\\/New folder\\\/wel...')\\n#8 \\\/home\\\/src\\\/owncloud\\\/server\\\/master\\\/3rdparty\\\/sabre\\\/dav\\\/lib\\\/DAV\\\/Tree.php(179): OCA\\\\DAV\\\\Connector\\\\Sabre\\\\File->delete()\\n#9 \\\/home\\\/src\\\/owncloud\\\/server\\\/master\\\/3rdparty\\\/sabre\\\/dav\\\/lib\\\/DAV\\\/CorePlugin.php(287): Sabre\\\\DAV\\\\Tree->delete('New folder\\\/welc...')\\n#10 [internal function]: Sabre\\\\DAV\\\\CorePlugin->httpDelete(Object(Sabre\\\\HTTP\\\\Request), Object(Sabre\\\\HTTP\\\\Response))\\n#11 \\\/home\\\/src\\\/owncloud\\\/server\\\/master\\\/3rdparty\\\/sabre\\\/event\\\/lib\\\/EventEmitterTrait.php(105): call_user_func_array(Array, Array)\\n#12 \\\/home\\\/src\\\/owncloud\\\/server\\\/master\\\/3rdparty\\\/sabre\\\/dav\\\/lib\\\/DAV\\\/Server.php(459): Sabre\\\\Event\\\\EventEmitter->emit('method:DELETE', Array)\\n#13 \\\/home\\\/src\\\/owncloud\\\/server\\\/master\\\/3rdparty\\\/sabre\\\/dav\\\/lib\\\/DAV\\\/Server.php(248): Sabre\\\\DAV\\\\Server->invokeMethod(Object(Sabre\\\\HTTP\\\\Request), Object(Sabre\\\\HTTP\\\\Response))\\n#14 \\\/home\\\/src\\\/owncloud\\\/server\\\/master\\\/apps\\\/dav\\\/appinfo\\\/v1\\\/webdav.php(54): Sabre\\\\DAV\\\\Server->exec()\\n#15 \\\/home\\\/src\\\/owncloud\\\/server\\\/master\\\/remote.php(137): require_once('\\\/home\\\/src\\\/owncl...')\\n#16 {main}\",\"File\":\"\\\/home\\\/src\\\/owncloud\\\/server\\\/master\\\/lib\\\/private\\\/files\\\/filesystem.php\",\"Line\":387}","level":3,"time":"2016-02-16T16:08:39+00:00"}
{"reqId":"h3FJVmACIo8+mISwiP+g","remoteAddr":"::1","app":"no app in context","message":"Exception: {\"Exception\":\"OCP\\\\Files\\\\NotFoundException\",\"Message\":\"File with id \\\"11\\\" has not been found.\",\"Code\":0,\"Trace\":\"#0 \\\/home\\\/src\\\/owncloud\\\/server\\\/master\\\/apps\\\/files_versions\\\/lib\\\/storage.php(98): OC\\\\Files\\\\View->getPath(11)\\n#1 \\\/home\\\/src\\\/owncloud\\\/server\\\/master\\\/apps\\\/files_versions\\\/lib\\\/storage.php(198): OCA\\\\Files_Versions\\\\Storage::getUidAndFilename('\\\/New folder\\\/wel...')\\n#2 \\\/home\\\/src\\\/owncloud\\\/server\\\/master\\\/apps\\\/files_versions\\\/lib\\\/hooks.php(90): OCA\\\\Files_Versions\\\\Storage::markDeletedFile('\\\/New folder\\\/wel...')\\n#3 [internal function]: OCA\\\\Files_Versions\\\\Hooks::pre_remove_hook(Array)\\n#4 \\\/home\\\/src\\\/owncloud\\\/server\\\/master\\\/lib\\\/private\\\/hook.php(104): call_user_func(Array, Array)\\n#5 \\\/home\\\/src\\\/owncloud\\\/server\\\/master\\\/lib\\\/private\\\/files\\\/view.php(1172): OC_Hook::emit('OC_Filesystem', 'delete', Array)\\n#6 \\\/home\\\/src\\\/owncloud\\\/server\\\/master\\\/lib\\\/private\\\/files\\\/view.php(1048): OC\\\\Files\\\\View->runHooks(Array, '\\\/New folder\\\/wel...')\\n#7 \\\/home\\\/src\\\/owncloud\\\/server\\\/master\\\/lib\\\/private\\\/files\\\/view.php(641): OC\\\\Files\\\\View->basicOperation('unlink', '\\\/New folder\\\/wel...', Array)\\n#8 \\\/home\\\/src\\\/owncloud\\\/server\\\/master\\\/apps\\\/dav\\\/lib\\\/connector\\\/sabre\\\/file.php(320): OC\\\\Files\\\\View->unlink('\\\/New folder\\\/wel...')\\n#9 \\\/home\\\/src\\\/owncloud\\\/server\\\/master\\\/3rdparty\\\/sabre\\\/dav\\\/lib\\\/DAV\\\/Tree.php(179): OCA\\\\DAV\\\\Connector\\\\Sabre\\\\File->delete()\\n#10 \\\/home\\\/src\\\/owncloud\\\/server\\\/master\\\/3rdparty\\\/sabre\\\/dav\\\/lib\\\/DAV\\\/CorePlugin.php(287): Sabre\\\\DAV\\\\Tree->delete('New folder\\\/welc...')\\n#11 [internal function]: Sabre\\\\DAV\\\\CorePlugin->httpDelete(Object(Sabre\\\\HTTP\\\\Request), Object(Sabre\\\\HTTP\\\\Response))\\n#12 \\\/home\\\/src\\\/owncloud\\\/server\\\/master\\\/3rdparty\\\/sabre\\\/event\\\/lib\\\/EventEmitterTrait.php(105): call_user_func_array(Array, Array)\\n#13 \\\/home\\\/src\\\/owncloud\\\/server\\\/master\\\/3rdparty\\\/sabre\\\/dav\\\/lib\\\/DAV\\\/Server.php(459): Sabre\\\\Event\\\\EventEmitter->emit('method:DELETE', Array)\\n#14 \\\/home\\\/src\\\/owncloud\\\/server\\\/master\\\/3rdparty\\\/sabre\\\/dav\\\/lib\\\/DAV\\\/Server.php(248): Sabre\\\\DAV\\\\Server->invokeMethod(Object(Sabre\\\\HTTP\\\\Request), Object(Sabre\\\\HTTP\\\\Response))\\n#15 \\\/home\\\/src\\\/owncloud\\\/server\\\/master\\\/apps\\\/dav\\\/appinfo\\\/v1\\\/webdav.php(54): Sabre\\\\DAV\\\\Server->exec()\\n#16 \\\/home\\\/src\\\/owncloud\\\/server\\\/master\\\/remote.php(137): require_once('\\\/home\\\/src\\\/owncl...')\\n#17 {main}\",\"File\":\"\\\/home\\\/src\\\/owncloud\\\/server\\\/master\\\/lib\\\/private\\\/files\\\/view.php\",\"Line\":1639}","level":3,"time":"2016-02-16T16:08:42+00:00"}
{"reqId":"h3FJVmACIo8+mISwiP+g","remoteAddr":"::1","app":"PHP","message":"Undefined index: \/New folder\/welcome.txt at \/home\/src\/owncloud\/server\/master\/apps\/files_versions\/lib\/storage.php#226","level":3,"time":"2016-02-16T16:08:45+00:00"}
```

@icewind1991 @PVince81 please have a look. I think that's the smallest possible change to fix it for 9.0. Not sure if we can do something else for 9.1 in order to use `$view->getUidAndFilename($filename);` again.

@rullzer I didn't revert the changes for apps/files_sharing/lib/helper.php yet: https://github.com/owncloud/core/commit/300eb54c871cfe48165ee32ecdc5067226aa0b7b#diff-407e98495650a16757767c1f2b423ddbR209
Do you think we should revert this too? Is it still used for sharing 2.0 and do you think it can create problems if the federated cloud ID is returned here instead of a local user?

fix https://github.com/owncloud/core/issues/22504

Commit:
get the actual user instead of a federated cloud id

$view->getUidAndFilename($filename); returns the federated cloud id in case of
a federated share. But in this case we need the local user who "owns" the file
which is the current logged in user in case of a federated share

PVince81 commented on apps/files_trashbin/lib/trashbin.php:
```diff
@@ -76,7 +76,22 @@ public static function ensureFileScannedHook($params) {
+		$uid = \OC\Files\Filesystem::getOwner($filename);
+		$userManager = \OC::$server->getUserManager();
+		if (!$userManager->userExists($uid)) {
+			$uid = \OCP\User::getUser();
```
Can you update the PHP doc to mention this behavior for both methods ?

PVince81 commented:
Not too happy, but I think it's fine as I can't think of a better way. :+1: 

Would be good to know what @icewind1991 thinks

schiesbn commented:
I updated the PR with a different approach. This way we don't need to have a copy of the same code in two apps and can re-use `$view->getUidAndFilename($filename);`

Please have another look, especially @icewind1991 at initMountPoints().

schiesbn commented:
@rullzer any idea why all the sharing tests fail? I don't see why this changes should have such a impact: https://github.com/owncloud/core/commit/48c4955722b2a8a45bf93a4fb48bf5be3cb6d4b5

rullzer commented:
@schiesbn probabaly some cyclic deps thing. If init mount points is called again (maybe from the sharing code) then https://github.com/owncloud/core/commit/48c4955722b2a8a45bf93a4fb48bf5be3cb6d4b5#diff-73d71f74c9f6d7093f369006c0c60dfeL375 won't just return now...

icewind1991 commented on lib/private/files/filesystem.php:
```diff
@@ -375,7 +375,6 @@ public static function initMountPoints($user = '') {
 		if (isset(self::$usersSetup[$user])) {
 			return;
 		}
-		self::$usersSetup[$user] = true;
```
This needs to be set early to prevent recursions in some cases

schiesbn replied:
Thanks, I will go back to the old approach

Commit:
get the actual user instead of a federated cloud id

$view->getUidAndFilename($filename); returns the federated cloud id in case of
a federated share. But in this case we need the local user who "owns" the file
which is the current logged in user in case of a federated share

schiesbn commented:
reverted back to the old approach. This works and is the minimal solution for 9.0 without any side-effects.

rullzer commented:
:+1: trashbin and versioning of federated shares work once again...

PVince81 commented:
:+1:

Dianafg76 commented:
:+1: Thanks :)
